### PR TITLE
Punishment Service implemented. Hooray.

### PR DESCRIPTION
**Pull request by Lamphobic**
PunishmentEntry made and is appropriately setup for ebeans and guice. PunishmentService made and setup for ebeans and guice. Logic migrated from BotService to appropriate locations. Punishment removed.

### CODE REVIEW

File: src/main/java/com/mattmerr/dr9k/service/BotService.java [hunk 1]
```diff
@@ -59,88 +54,35 @@ public void disconnect() {
...
-        System.out.println("Removing message from user.");
-        event.getMessage().delete().queue();
-        user.openPrivateChannel().queue((channel) -> channel.sendMessage(String
-            .format(
-                "You are currently muted in %s. Please wait until your mute "
-                    + "ends in %s to try again.",
-                location, getPunishment(user).getHumanTimeRemaining()))
-            .queue());
+    if(userHasPunishmentRecord(message)){
+      if(punishmentService.handlePunishment(message)){
```
Comment (MatthewMerrill) [inline src/main/java/com/mattmerr/dr9k/service/BotService.java]: return value of handlePunishment is unclear, could you add an intermediate variable for clarity

```
// (or wasPunished or punishmentExtended or whatever's best name)
boolean punishedForMessage = handlePunishment(message)
```

File: src/main/java/com/mattmerr/dr9k/service/PunishmentService.java [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+        .eq("authorId", message.getAuthor().getId()).findOne();
+  }
+
+  public int removePunishment(@NotNull Message message){
+    return db.createQuery(PunishmentEntry.class).where()
+        .eq("guildId", message.getGuild().getId())
+        .eq("authorId", message.getAuthor().getId()).delete();
+  }
+
+  public boolean handlePunishment(Message message) {
```
Comment (MatthewMerrill) [inline src/main/java/com/mattmerr/dr9k/service/PunishmentService.java]: Maybe handlePunishment could be renamed "enforcePunishment"

File: src/main/java/com/mattmerr/dr9k/service/PunishmentService.java [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+    PunishmentEntry userPunishment = getPunishment(message);
+    String location = String.format("Guild Server: %s, Channel: %s",
+        message.getGuild().getName(),
+        message.getChannel().getName());
+    System.out.println("Punished user detected: " + message.getAuthor().getName());
+    if (userPunishment.isOver()) {
+      System.out.println("User is not currently muted.");
+      if (userPunishment.getPunishmentDecayed()) {
+        removePunishmentRecord(message);
+      }
```
Comment (MatthewMerrill) [inline src/main/java/com/mattmerr/dr9k/service/PunishmentService.java]: `return false` here and remove the else branch?

File: src/main/java/com/mattmerr/dr9k/model/BlacklistEntry.java [hunk 4]
```diff
@@ -12,6 +12,7 @@
 @Entity
 @Table(name = "blacklist_entries")
 public class BlacklistEntry extends Model {
+  private final int decaySeconds = 30; //21600 for 6 hours
```
[the remark this answers is not in the document]
Comment (Lamphobic) [reply src/main/java/com/mattmerr/dr9k/model/BlacklistEntry.java]: Oops I literally have no idea how that got there.

File: src/main/java/com/mattmerr/dr9k/model/PunishmentEntry.java [hunk 5]
```diff
@@ -1,21 +1,104 @@
...
-  private final int decaySeconds = 30; //21600 for 6 hours
+  @Id
+  private Long id;
 
-  public Punishment(int level, Instant startTime) {
-    this.severityLevel = level;
-    this.punishmentStart = startTime;
+  @NotNull
+  @Length(512)
+  private String guildId;
```
[the remark this answers is not in the document]
Comment (Lamphobic) [reply src/main/java/com/mattmerr/dr9k/model/PunishmentEntry.java]: That's next feature. Putting the TODO at the botservice inject.

File: src/main/java/com/mattmerr/dr9k/service/PunishmentService.java (hunk 2 above)
Comment (Lamphobic) [reply src/main/java/com/mattmerr/dr9k/service/PunishmentService.java]: I like the rename on that. Will do.

File: src/main/java/com/mattmerr/dr9k/model/PunishmentEntry.java [hunk 6]
```diff
@@ -1,21 +1,104 @@
...
+import javax.persistence.Table;
 import java.time.Duration;
 import java.time.Instant;
 import java.time.temporal.ChronoUnit;
-import java.util.Objects;
 
-public class Punishment {
+@Entity
+@Table(name = "punishment_entries")
+public class PunishmentEntry {
```
[the remark this answers is not in the document]
Comment (Lamphobic) [reply src/main/java/com/mattmerr/dr9k/model/PunishmentEntry.java]: I assume you're asking me to rename the class and table name?

**Comment by MatthewMerrill** [comment]
Yeah, IMO the "Entry" suffix is necessary for describing an entry in the
column of blacklisted items bc blacklist is a collection but the table of
punishments can just be called punishments

File: src/main/java/com/mattmerr/dr9k/service/PunishmentService.java (hunk 3 above)
Comment (Lamphobic) [reply src/main/java/com/mattmerr/dr9k/service/PunishmentService.java]: Wouldn't work. This chungus could be broken down better. I'm handling three cases here. Reminder that this is only called if there exists a record of punishment in the db. In order of appearance in code: User is no longer muted, but severity has not decayed enough to remove the  record. User is no longer muted and seveer has decayed enough to remove the record, user is still muted. See new changes. They'll make more sense.

File: src/main/java/com/mattmerr/dr9k/model/Punishment.java [hunk 7]
```diff
@@ -1,21 +1,104 @@
...
+  }
+
+  public int getPunishmentDecay() {
+    Instant now = Instant.now();
+    Duration timePassed = Duration.between(punishmentStart, now);
+    long secondsPassed = timePassed.getSeconds();
+    return (int) secondsPassed / PunishmentService.getDecaySeconds();
   }
 
   public String getHumanTimeRemaining() {
```
Comment (MatthewMerrill) [inline src/main/java/com/mattmerr/dr9k/model/Punishment.java]: nit: is it possible to do this with SimpleDateFormat or something? Maybe just add `// TODO: replace with date formatter`

Comment (Lamphobic) [reply src/main/java/com/mattmerr/dr9k/model/Punishment.java]: Timezones.

Comment (Lamphobic) [reply src/main/java/com/mattmerr/dr9k/model/Punishment.java]: I considered at one point using DateTime at GMT+0 as my core storage solution and returning an end-time at GMT+0, but decided this was better. If you want to argue for it though, go ahead.

**Pull request merged**